core-services/workflow/src/main/java/org/silverpeas/core/workflow/engine/instance/ProcessInstanceImpl.java

```diff
@@ -1771,67 +1544,49 @@ public void setTimeoutStatusCastor(int timeoutStatus) {
   private void addUndoHistoryStep(String action, String params) throws WorkflowException {
     UndoHistoryStep undoStep = new UndoHistoryStep();
     undoStep.setStepId(this.currentStep.getId());
-    undoStep.setInstanceId(instanceId);
+    undoStep.setInstance(this);
     undoStep.setAction(action);
     undoStep.setParameters(params);
 
-    Database db = null;
-    try {
-      db = WorkflowJDOManager.getDatabase();
-      db.begin();
-      db.create(undoStep);
-      db.commit();
-    } catch (PersistenceException pe) {
-      throw new WorkflowException("ProcessInstanceImpl.addUndoHistoryStep",
-          "workflowEngine.EX_ERR_CASTOR_CREATE_UNDO_HISTORYSTEP", "instanceid=" + getInstanceId(),
-          pe);
-    } finally {
-      WorkflowJDOManager.closeDatabase(db);
+    undoSteps.add(undoStep);
+  }
+
+  private List<UndoHistoryStep> getByStepId(String id) {
+    List<UndoHistoryStep> undoStepsOfStep = new ArrayList<>();
+    for (UndoHistoryStep undoStep : undoSteps) {
+      if (undoStep.getId().equals(id)) {
+        undoStepsOfStep.add(undoStep);
+      }
     }
+    return undoStepsOfStep;
   }
 
   /**
    * Undo all atomic operations that had occured for a given historyStep
    * @param historyStep the historyStep when the atomic operations had occured
    */
   private void undoStep(HistoryStep historyStep) throws WorkflowException {
-    OQLQuery query = null;
-    QueryResults results;
-    Database db = null;
-
     try {
       // Mark this instance as beeing in undo process
       // to avoid storing atomic operation done here
       this.inUndoProcess = true;
 
-      // Constructs the query
-      db = WorkflowJDOManager.getDatabase();
-      db.begin();
-      query = db.getOQLQuery(
-          "SELECT undoStep " +
-          "FROM org.silverpeas.core.workflow.engine.instance.UndoHistoryStep undoStep " +
-          "WHERE undoStep.stepId = $1 ");
-
-      // Execute the query
-      query.bind((Integer.parseInt(historyStep.getId())));
-      results = query.execute();
-
-      while (results.hasMore()) {
-        UndoHistoryStep undoStep = (UndoHistoryStep) results.next();
+      List<UndoHistoryStep> undoSteps = getByStepId(historyStep.getId());
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Rename "undoSteps" which hides the field declared at line 121. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AHiddenFieldCheck)